Discriminated union accessed via 'this' can have an incorrect type in subsequent uses

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**
3.1.3
3.5.2
3.6.0-dev.20190619
2.0.10

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
discriminated union this class method duplicate name TS2322

**Code**

```ts
type State = {
    type: "numberVariant";
    data: number;
} | {
    type: "stringVariant";
    data: string;
};

class SomeClass {
    state: State;
    method() {
        while (0) { }

        this.state.data;
        
        if (this.state.type === "stringVariant") {
            let s: string = this.state.data;
        }
    }
}
```

**Expected behavior:**
There should be no error.

**Actual behavior:**
ts-bug.ts:17:17 - error TS2322: Type 'string | number' is not assignable to type 'string'.

All the types look correct on hover (on the line with the error, hovering over `data` shows `(property) data: string`), but there's an error anyway.  It seems all subsequent uses of `this.state.data` use the type of its first use, e.g. if you wrap the first `this.state.data` in a check for numberVariant, it will change the type in the error to 'number', and [will even mistakenly let you use it as a number outside of the check](https://www.typescriptlang.org/play/#src=type%20State%20%3D%20%7B%0D%0A%20%20%20%20type%3A%20%22numberVariant%22%3B%0D%0A%20%20%20%20data%3A%20number%3B%0D%0A%7D%20%7C%20%7B%0D%0A%20%20%20%20type%3A%20%22stringVariant%22%3B%0D%0A%20%20%20%20data%3A%20string%3B%0D%0A%7D%3B%0D%0A%0D%0Aclass%20SomeClass%20%7B%0D%0A%20%20%20%20state%3A%20State%3B%0D%0A%20%20%20%20method()%20%7B%0D%0A%20%20%20%20%20%20%20%20while%20(0)%20%7B%20%7D%0D%0A%0D%0A%20%20%20%20%20%20%20%20if%20(this.state.type%20%3D%3D%3D%20%22numberVariant%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20this.state.data%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20%0D%0A%20%20%20%20%20%20%20%20let%20n%3A%20number%20%3D%20this.state.data%3B%20%2F%2F%20This%20should%20be%20an%20error%0D%0A%20%20%20%20%7D%0D%0A%7D).

Commenting out the loop or the first usage of `this.state.data` makes the error go away.  It also works correctly when doing `const that = this;` at the start of the method and using `that` everywhere instead.  Not reusing property names for the different variants also avoids this issue, so it's easy to workaround.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->
[Link](https://www.typescriptlang.org/play/#src=type%20State%20%3D%20%7B%0D%0A%20%20%20%20type%3A%20%22numberVariant%22%3B%0D%0A%20%20%20%20data%3A%20number%3B%0D%0A%7D%20%7C%20%7B%0D%0A%20%20%20%20type%3A%20%22stringVariant%22%3B%0D%0A%20%20%20%20data%3A%20string%3B%0D%0A%7D%3B%0D%0A%0D%0Aclass%20SomeClass%20%7B%0D%0A%20%20%20%20state%3A%20State%3B%0D%0A%20%20%20%20method()%20%7B%0D%0A%20%20%20%20%20%20%20%20while%20(0)%20%7B%20%7D%0D%0A%0D%0A%20%20%20%20%20%20%20%20this.state.data%3B%0D%0A%20%20%20%20%20%20%20%20%0D%0A%20%20%20%20%20%20%20%20if%20(this.state.type%20%3D%3D%3D%20%22stringVariant%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20let%20s%3A%20string%20%3D%20this.state.data%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%7D)

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
This has apparently been the case since 2.0, so there's probably issues already but I can't find them.  But between the loop, `this`, and reused property name it reminds me a lot of #14860, which was apparently some sort of caching issue.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -21370,7 +21370,7 @@ namespace ts {
                 // The result is incomplete if the first antecedent (the non-looping control flow path)
                 // is incomplete.
                 const result = getUnionOrEvolvingArrayType(antecedentTypes, subtypeReduction ? UnionReduction.Subtype : UnionReduction.Literal);
-                 if (isIncomplete(firstAntecedentType!)) {
+                if (isIncomplete(firstAntecedentType!)) {
                     return createFlowType(result, /*incomplete*/ true);
                 }
                 cache.set(key, result);
@@ -30030,19 +30030,19 @@ namespace ts {
                 return quickType;
             }
             // If a type has been cached for the node, return it.
-            // if (node.flags & NodeFlags.TypeCached && flowTypeCache) {
-            //     const cachedType = flowTypeCache[getNodeId(node)];
-            //     if (cachedType) {
-            //         return cachedType;
-            //     }
-            // }
+            if (node.flags & NodeFlags.TypeCached && flowTypeCache) {
+                const cachedType = flowTypeCache[getNodeId(node)];
+                if (cachedType) {
+                    return cachedType;
+                }
+            }
             const startInvocationCount = flowInvocationCount;
             const type = checkExpression(node);
             // If control flow analysis was required to determine the type, it is worth caching.
             if (flowInvocationCount !== startInvocationCount) {
-            //     const cache = flowTypeCache || (flowTypeCache = []);
-            //     cache[getNodeId(node)] = type;
-            //     setNodeFlags(node, node.flags | NodeFlags.TypeCached);
+                const cache = flowTypeCache || (flowTypeCache = []);
+                cache[getNodeId(node)] = type;
+                setNodeFlags(node, node.flags | NodeFlags.TypeCached);
             }
             return type;
         }

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -20078,7 +20078,7 @@ namespace ts {
                     const symbol = getResolvedSymbol(<Identifier>node);
                     return symbol !== unknownSymbol ? `${flowContainer ? getNodeId(flowContainer) : "-1"}|${getTypeId(declaredType)}|${getTypeId(initialType)}|${isConstraintPosition(node) ? "@" : ""}${getSymbolId(symbol)}` : undefined;
                 case SyntaxKind.ThisKeyword:
-                    return "0";
+                    return `0|${flowContainer ? getNodeId(flowContainer) : "-1"}|${getTypeId(declaredType)}|${getTypeId(initialType)}`;
                 case SyntaxKind.NonNullExpression:
                 case SyntaxKind.ParenthesizedExpression:
                     return getFlowCacheKey((<NonNullExpression | ParenthesizedExpression>node).expression, declaredType, initialType, flowContainer);
@@ -20941,7 +20941,7 @@ namespace ts {
 
         function getFlowTypeOfReference(reference: Node, declaredType: Type, initialType = declaredType, flowContainer?: Node, couldBeUninitialized?: boolean) {
             let key: string | undefined;
-            let keySet = false;
+            let isKeySet = false;
             let flowDepth = 0;
             if (flowAnalysisDisabled) {
                 return errorType;
@@ -20964,10 +20964,10 @@ namespace ts {
             return resultType;
 
             function getOrSetCacheKey() {
-                if (keySet) {
+                if (isKeySet) {
                     return key;
                 }
-                keySet = true;
+                isKeySet = true;
                 return key = getFlowCacheKey(reference, declaredType, initialType, flowContainer);
             }
 
@@ -21370,7 +21370,7 @@ namespace ts {
                 // The result is incomplete if the first antecedent (the non-looping control flow path)
                 // is incomplete.
                 const result = getUnionOrEvolvingArrayType(antecedentTypes, subtypeReduction ? UnionReduction.Subtype : UnionReduction.Literal);
-                if (isIncomplete(firstAntecedentType!)) {
+                 if (isIncomplete(firstAntecedentType!)) {
                     return createFlowType(result, /*incomplete*/ true);
                 }
                 cache.set(key, result);
@@ -30030,19 +30030,19 @@ namespace ts {
                 return quickType;
             }
             // If a type has been cached for the node, return it.
-            if (node.flags & NodeFlags.TypeCached && flowTypeCache) {
-                const cachedType = flowTypeCache[getNodeId(node)];
-                if (cachedType) {
-                    return cachedType;
-                }
-            }
+            // if (node.flags & NodeFlags.TypeCached && flowTypeCache) {
+            //     const cachedType = flowTypeCache[getNodeId(node)];
+            //     if (cachedType) {
+            //         return cachedType;
+            //     }
+            // }
             const startInvocationCount = flowInvocationCount;
             const type = checkExpression(node);
             // If control flow analysis was required to determine the type, it is worth caching.
             if (flowInvocationCount !== startInvocationCount) {
-                const cache = flowTypeCache || (flowTypeCache = []);
-                cache[getNodeId(node)] = type;
-                setNodeFlags(node, node.flags | NodeFlags.TypeCached);
+            //     const cache = flowTypeCache || (flowTypeCache = []);
+            //     cache[getNodeId(node)] = type;
+            //     setNodeFlags(node, node.flags | NodeFlags.TypeCached);
             }
             return type;
         }

--- a/tests/baselines/reference/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.errors.txt
+++ b/tests/baselines/reference/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.errors.txt
@@ -1,0 +1,39 @@
+tests/cases/conformance/classes/members/instanceAndStaticMembers/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts(29,13): error TS2322: Type 'string | number' is not assignable to type 'number'.
+  Type 'string' is not assignable to type 'number'.
+
+
+==== tests/cases/conformance/classes/members/instanceAndStaticMembers/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts (1 errors) ====
+    // #31995
+    type State = {
+        type: "numberVariant";
+        data: number;
+    } | {
+        type: "stringVariant";
+        data: string;
+    };
+    
+    class SomeClass {
+        state!: State;
+        method() {
+            while (0) { }
+            this.state.data;
+            if (this.state.type === "stringVariant") {
+                const s: string = this.state.data;
+            }
+        }
+    }
+    
+    class SomeClass2 {
+        state!: State;
+        method() {
+            const c = false;
+            while (c) { }
+            if (this.state.type === "numberVariant") {
+                this.state.data;
+            }
+            let n: number = this.state?.data; // This should be an error
+                ~
+!!! error TS2322: Type 'string | number' is not assignable to type 'number'.
+!!! error TS2322:   Type 'string' is not assignable to type 'number'.
+        }
+    }

--- a/tests/baselines/reference/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.js
+++ b/tests/baselines/reference/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.js
@@ -1,0 +1,60 @@
+//// [typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts]
+// #31995
+type State = {
+    type: "numberVariant";
+    data: number;
+} | {
+    type: "stringVariant";
+    data: string;
+};
+
+class SomeClass {
+    state!: State;
+    method() {
+        while (0) { }
+        this.state.data;
+        if (this.state.type === "stringVariant") {
+            const s: string = this.state.data;
+        }
+    }
+}
+
+class SomeClass2 {
+    state!: State;
+    method() {
+        const c = false;
+        while (c) { }
+        if (this.state.type === "numberVariant") {
+            this.state.data;
+        }
+        let n: number = this.state?.data; // This should be an error
+    }
+}
+
+//// [typeOfThisInstanceMemberNarrowedWithLoopAntecedent.js]
+var SomeClass = /** @class */ (function () {
+    function SomeClass() {
+    }
+    SomeClass.prototype.method = function () {
+        while (0) { }
+        this.state.data;
+        if (this.state.type === "stringVariant") {
+            var s = this.state.data;
+        }
+    };
+    return SomeClass;
+}());
+var SomeClass2 = /** @class */ (function () {
+    function SomeClass2() {
+    }
+    SomeClass2.prototype.method = function () {
+        var _a;
+        var c = false;
+        while (c) { }
+        if (this.state.type === "numberVariant") {
+            this.state.data;
+        }
+        var n = (_a = this.state) === null || _a === void 0 ? void 0 : _a.data; // This should be an error
+    };
+    return SomeClass2;
+}());

--- a/tests/baselines/reference/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.symbols
+++ b/tests/baselines/reference/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.symbols
@@ -1,0 +1,95 @@
+=== tests/cases/conformance/classes/members/instanceAndStaticMembers/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts ===
+// #31995
+type State = {
+>State : Symbol(State, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 0, 0))
+
+    type: "numberVariant";
+>type : Symbol(type, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 1, 14))
+
+    data: number;
+>data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 2, 26))
+
+} | {
+    type: "stringVariant";
+>type : Symbol(type, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 4, 5))
+
+    data: string;
+>data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 5, 26))
+
+};
+
+class SomeClass {
+>SomeClass : Symbol(SomeClass, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 7, 2))
+
+    state!: State;
+>state : Symbol(SomeClass.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 9, 17))
+>State : Symbol(State, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 0, 0))
+
+    method() {
+>method : Symbol(SomeClass.method, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 10, 18))
+
+        while (0) { }
+        this.state.data;
+>this.state.data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 2, 26), Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 5, 26))
+>this.state : Symbol(SomeClass.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 9, 17))
+>this : Symbol(SomeClass, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 7, 2))
+>state : Symbol(SomeClass.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 9, 17))
+>data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 2, 26), Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 5, 26))
+
+        if (this.state.type === "stringVariant") {
+>this.state.type : Symbol(type, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 1, 14), Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 4, 5))
+>this.state : Symbol(SomeClass.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 9, 17))
+>this : Symbol(SomeClass, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 7, 2))
+>state : Symbol(SomeClass.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 9, 17))
+>type : Symbol(type, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 1, 14), Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 4, 5))
+
+            const s: string = this.state.data;
+>s : Symbol(s, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 15, 17))
+>this.state.data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 5, 26))
+>this.state : Symbol(SomeClass.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 9, 17))
+>this : Symbol(SomeClass, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 7, 2))
+>state : Symbol(SomeClass.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 9, 17))
+>data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 5, 26))
+        }
+    }
+}
+
+class SomeClass2 {
+>SomeClass2 : Symbol(SomeClass2, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 18, 1))
+
+    state!: State;
+>state : Symbol(SomeClass2.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 20, 18))
+>State : Symbol(State, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 0, 0))
+
+    method() {
+>method : Symbol(SomeClass2.method, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 21, 18))
+
+        const c = false;
+>c : Symbol(c, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 23, 13))
+
+        while (c) { }
+>c : Symbol(c, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 23, 13))
+
+        if (this.state.type === "numberVariant") {
+>this.state.type : Symbol(type, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 1, 14), Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 4, 5))
+>this.state : Symbol(SomeClass2.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 20, 18))
+>this : Symbol(SomeClass2, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 18, 1))
+>state : Symbol(SomeClass2.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 20, 18))
+>type : Symbol(type, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 1, 14), Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 4, 5))
+
+            this.state.data;
+>this.state.data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 2, 26))
+>this.state : Symbol(SomeClass2.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 20, 18))
+>this : Symbol(SomeClass2, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 18, 1))
+>state : Symbol(SomeClass2.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 20, 18))
+>data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 2, 26))
+        }
+        let n: number = this.state?.data; // This should be an error
+>n : Symbol(n, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 28, 11))
+>this.state?.data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 2, 26), Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 5, 26))
+>this.state : Symbol(SomeClass2.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 20, 18))
+>this : Symbol(SomeClass2, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 18, 1))
+>state : Symbol(SomeClass2.state, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 20, 18))
+>data : Symbol(data, Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 2, 26), Decl(typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts, 5, 26))
+    }
+}

--- a/tests/baselines/reference/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.types
+++ b/tests/baselines/reference/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.types
@@ -1,0 +1,100 @@
+=== tests/cases/conformance/classes/members/instanceAndStaticMembers/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts ===
+// #31995
+type State = {
+>State : State
+
+    type: "numberVariant";
+>type : "numberVariant"
+
+    data: number;
+>data : number
+
+} | {
+    type: "stringVariant";
+>type : "stringVariant"
+
+    data: string;
+>data : string
+
+};
+
+class SomeClass {
+>SomeClass : SomeClass
+
+    state!: State;
+>state : State
+
+    method() {
+>method : () => void
+
+        while (0) { }
+>0 : 0
+
+        this.state.data;
+>this.state.data : string | number
+>this.state : State
+>this : this
+>state : State
+>data : string | number
+
+        if (this.state.type === "stringVariant") {
+>this.state.type === "stringVariant" : boolean
+>this.state.type : "numberVariant" | "stringVariant"
+>this.state : State
+>this : this
+>state : State
+>type : "numberVariant" | "stringVariant"
+>"stringVariant" : "stringVariant"
+
+            const s: string = this.state.data;
+>s : string
+>this.state.data : string
+>this.state : { type: "stringVariant"; data: string; }
+>this : this
+>state : { type: "stringVariant"; data: string; }
+>data : string
+        }
+    }
+}
+
+class SomeClass2 {
+>SomeClass2 : SomeClass2
+
+    state!: State;
+>state : State
+
+    method() {
+>method : () => void
+
+        const c = false;
+>c : false
+>false : false
+
+        while (c) { }
+>c : false
+
+        if (this.state.type === "numberVariant") {
+>this.state.type === "numberVariant" : boolean
+>this.state.type : "numberVariant" | "stringVariant"
+>this.state : State
+>this : this
+>state : State
+>type : "numberVariant" | "stringVariant"
+>"numberVariant" : "numberVariant"
+
+            this.state.data;
+>this.state.data : number
+>this.state : { type: "numberVariant"; data: number; }
+>this : this
+>state : { type: "numberVariant"; data: number; }
+>data : number
+        }
+        let n: number = this.state?.data; // This should be an error
+>n : number
+>this.state?.data : string | number
+>this.state : State
+>this : this
+>state : State
+>data : string | number
+    }
+}

--- a/tests/cases/conformance/classes/members/instanceAndStaticMembers/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts
+++ b/tests/cases/conformance/classes/members/instanceAndStaticMembers/typeOfThisInstanceMemberNarrowedWithLoopAntecedent.ts
@@ -1,0 +1,31 @@
+// #31995
+type State = {
+    type: "numberVariant";
+    data: number;
+} | {
+    type: "stringVariant";
+    data: string;
+};
+
+class SomeClass {
+    state!: State;
+    method() {
+        while (0) { }
+        this.state.data;
+        if (this.state.type === "stringVariant") {
+            const s: string = this.state.data;
+        }
+    }
+}
+
+class SomeClass2 {
+    state!: State;
+    method() {
+        const c = false;
+        while (c) { }
+        if (this.state.type === "numberVariant") {
+            this.state.data;
+        }
+        let n: number = this.state?.data; // This should be an error
+    }
+}

